### PR TITLE
Revised calculation of wind GUST for UFS applications

### LINKED ISSUE #636: Gust wind weaker than mean wind in GFS
It was reported from the weather centers that sometimes the GFS model wind gusts in GRIB files are weaker than the mean wind, especially in weak wind situations. And in general, the gusts of wind are often underestimated compared to other models.

I checked the real-time Ops GFS and also my outputs from GFS.v16.3.4 tests (ICs: 2020082300).
I did see that sometimes the GFS model wind gusts in GRIB files (GUST 180) are weaker than
the mean wind, especially during daytime.

I checked the code, "gfs_ncep_post.fd" and "CALGUST.f", which computes surface wind gust by mixing
down momentum from the level at the PBL height. My understanding is that the GUST is derived
from the winds at the PBL height, U & V(I, J, LPBL). So, sometimes, GUST is smaller than the Wind@10m.

Solution:  
We are trying one solution and using the GSL algorithm for GUST calculation in GFS (UPP).  And if this solution works well, we will also update the GUST in HAFS and RRFS.

### PR DESCRIPTION
**Pull request by wzzheng90**
This PR is to address the issue "Gust wind weaker than mean wind in GFS" #636 (https://github.com/NOAA-EMC/UPP/issues/636).

### CODE REVIEW

File: sorc/ncep_post.fd/CALGUST.f [hunk 1]
```diff
@@ -133,7 +133,7 @@ SUBROUTINE CALGUST(LPBL,ZPBL,GUST)
              else
                sfcwind = spval
              endif
-             if(MODELNAME == 'RAPR') then
+             if(MODELNAME == 'RAPR' .OR. MODELNAME == 'GFS' .OR. MODELNAME == 'FV3R')) then
```
Comment (WenMeng-NOAA) [inline sorc/ncep_post.fd/CALGUST.f]: @wzzheng90 Please correct syntax error in your changes.

**Comment by WenMeng-NOAA** [comment]
@wzzheng90 I would suggest to add the following changes:

- In CALGUST.f: add change log with your name after line 18
- In CALGUST.f :  line 78-79 
   change
      CALL EXCH(UH(1,jsta_2l,L))
      CALL EXCH(VH(1,jsta_2l,L))
  into
      CALL EXCH(UH(ista_2l,jsta_2l,L))
      CALL EXCH(VH(ista_2l,jsta_2l,L))
  These changes are for 2D decomposition update and will have no impact on results you finished.

- In MDLFLD.f: line 3777
   Change
       IF(MODELNAME  ==  'GFS')THEN
    into
       IF(MODELNAME == 'GFS' .OR. MODELNAME == 'FV3R')THEN
   Also add a change log in this subroutine.

**Comment by wzzheng90** [comment]
> @wzzheng90 I would suggest to add the following changes:
> 
>     * In CALGUST.f: add change log with your name after line 18
> 
>     * In CALGUST.f :  line 78-79
>       change
>       CALL EXCH(UH(1,jsta_2l,L))
>       CALL EXCH(VH(1,jsta_2l,L))
>       into
>       CALL EXCH(UH(ista_2l,jsta_2l,L))
>       CALL EXCH(VH(ista_2l,jsta_2l,L))
>       These changes are for 2D decomposition update and will have no impact on results you finished.
> 
>     * In MDLFLD.f: line 3777
>       Change
>       IF(MODELNAME  ==  'GFS')THEN
>       into
>       IF(MODELNAME == 'GFS' .OR. MODELNAME == 'FV3R')THEN
>       Also add a change log in this subroutine.

I added your suggestions for 1) and 2). Thanks!!    
However, I am reluctant to add the 3rd change.  In fact, in "MDLFLD.f", Line 3777, it needs to compute PBL height based on Richardson number.   GFS does use Richardson number to calculate the PBL height (PBLRI) so we can directly use the PBL height from GFS moel.  But for other models such as HAFS or RRFS, it needs to re-calculate PBLRI.   I did a test and the result is quite different if using the PBL height from HAFS or RRFS.

**Comment by WenMeng-NOAA** [comment]
@wzzheng90 

> However, I am reluctant to add the 3rd change. In fact, in "MDLFLD.f", Line 3777, it needs to compute PBL height based on Richardson number. GFS does use Richardson number to calculate the PBL height (PBLRI) so we can directly use the PBL height from GFS moel. But for other models such as HAFS or RRFS, it needs to re-calculate PBLRI. I did a test and the result is quite different if using the PBL height from HAFS or RRFS.

GFS output PBL height as "hpbl" in netcdf. The UPP read it as pblh (see [here](https://github.com/NOAA-EMC/UPP/blob/develop/sorc/ncep_post.fd/INITPOST_NETCDF.f#L1620)) The line 3777 is for assigning gfs's pblh as pblri. I checked rrfs and hafs model history files. The hpbl is included. So I might think the modelname "FV3R" should be added in line 3777 to keep all ufs models consistent.

**Comment by wzzheng90** [comment]
> @wzzheng90
> 
> > However, I am reluctant to add the 3rd change. In fact, in "MDLFLD.f", Line 3777, it needs to compute PBL height based on Richardson number. GFS does use Richardson number to calculate the PBL height (PBLRI) so we can directly use the PBL height from GFS moel. But for other models such as HAFS or RRFS, it needs to re-calculate PBLRI. I did a test and the result is quite different if using the PBL height from HAFS or RRFS.
> 
> GFS output PBL height as "hpbl" in netcdf. The UPP read it as pblh (see [here](https://github.com/NOAA-EMC/UPP/blob/develop/sorc/ncep_post.fd/INITPOST_NETCDF.f#L1620)) The line 3777 is for assigning gfs's pblh as pblri. I checked rrfs and hafs model history files. The hpbl is included. So I might think the modelname "FV3R" should be added in line 3777 to keep all ufs models consistent.

Maybe I did not explain it clearly.  My point is that the PBL height from RRFS or HAFS model would not be derived based on Richardson number (different PBL schemes), so here it needs to re-calculate.  I am not clear why here it needs Ri-based PBL height.
So this would be discussed later. Anyway, let's include "FV3R" at present.  Thanks.

**Comment by WenMeng-NOAA** [comment]
> > @wzzheng90
> > > However, I am reluctant to add the 3rd change. In fact, in "MDLFLD.f", Line 3777, it needs to compute PBL height based on Richardson number. GFS does use Richardson number to calculate the PBL height (PBLRI) so we can directly use the PBL height from GFS moel. But for other models such as HAFS or RRFS, it needs to re-calculate PBLRI. I did a test and the result is quite different if using the PBL height from HAFS or RRFS.
> > 
> > 
> > GFS output PBL height as "hpbl" in netcdf. The UPP read it as pblh (see [here](https://github.com/NOAA-EMC/UPP/blob/develop/sorc/ncep_post.fd/INITPOST_NETCDF.f#L1620)) The line 3777 is for assigning gfs's pblh as pblri. I checked rrfs and hafs model history files. The hpbl is included. So I might think the modelname "FV3R" should be added in line 3777 to keep all ufs models consistent.
> 
> Maybe I did not explain it clearly. My point is that the PBL height from RRFS or HAFS model would not be derived based on Richardson number (different PBL schemes), so here it needs to re-calculate. I am not clear why here it needs Ri-based PBL height. So this would be discussed later. Anyway, let's include "FV3R" at present. Thanks.

@wzzheng90 Are you meaning at model side, global FV3 and regional FV3 use different schemes to calculate PBL height? In that case,  there is no need to add change in line 3777.

**Comment by wzzheng90** [comment]
@WenMeng-NOAA  yes, using different PBL schemes.  However, it may need to evaluate the results against the Obs to see which one is better in the future.

**Comment by WenMeng-NOAA** [comment]
The UPP RT tests were completed on WCOSS2, Hera and Orion. The baselines for gfs and hafs and rrfs will be recreated due to gust calculation update.
```
gfs: 
    master--
   14:21345016:GUST:surface:rpn_corr=0.996106:rpn_rms=0.706175
hafs: 
    HURPRS--
   3:2977237:GUST:surface:rpn_corr=0.994846:rpn_rms=0.556539
rrfs:  
    NATLEV--
   1286:5456532867:GUST:surface:rpn_corr=0.99558:rpn_rms=0.698341
    PRSLEV--
    51:293497035:GUST:surface:rpn_corr=0.99558:rpn_rms=0.698341   
```

**Review by WenMeng-NOAA — APPROVED** [review]

**Comment by WenMeng-NOAA** [comment]
I will proceed to merging this PR and skip the CI failure from intel build. That failure have no impact on UPP performance on WCOSS2 and other HPC platforms. The fix could be implemented later.

**Pull request merged**